File: processador_dados.py
import pandas as pd
import re
from calculadora_sla import CalculadoraSLA

def identificar_empresa(service_str):
    if not isinstance(service_str, str): return "Outros"
    s = service_str.upper()
    if "CEARA" in s or "COELCE" in s: return "CE"
    if "RIO" in s or "AMPLA" in s: return "RJ"
    if "SAO PAULO" in s or "SP" in s: return "SP"
    return "Outros"

def extrair_partes_description(short_desc):
    if not isinstance(short_desc, str):
        return "N/A", "N/A", "N/A", "N/A"
    
    macro, categoria, subcategoria, desc_limpa = "N/A", "N/A", "N/A", short_desc
    
    macro_match = re.search(r'\[(.*?)\]', short_desc)
    if macro_match:
        macro = macro_match.group(1).strip().upper()
        desc_limpa = desc_limpa.replace(macro_match.group(0), "")
        
    cat_match = re.search(r'\((.*?)\)', desc_limpa)
    if cat_match:
        conteudo = cat_match.group(1)
        partes = conteudo.split('/')
        categoria = partes[0].strip().upper() if len(partes) > 0 else "N/A"
        subcategoria = partes[1].strip().upper() if len(partes) > 1 else "N/A"
        desc_limpa = desc_limpa.replace(cat_match.group(0), "")
    
    desc_limpa = re.sub(r'\s+', ' ', desc_limpa).strip()
    if not desc_limpa: desc_limpa = "N/A"
        
    return macro, categoria, subcategoria, desc_limpa

def formatar_mes_pt(data):
    meses = ['Jan', 'Fev', 'Mar', 'Abr', 'Mai', 'Jun', 'Jul', 'Ago', 'Set', 'Out', 'Nov', 'Dez']
    if pd.isna(data): return "N/A"
    return f"{meses[data.month - 1]}/{str(data.year)[2:]}"

# Exportações "cruas" do ServiceNow usam nomes de coluna técnicos (snake_case).
# Mapeia para o formato interno esperado pelo app quando a coluna "bonita"
# (ex: "Opened") não estiver presente no arquivo importado.
_MAPA_COLUNAS_SERVICENOW = {
    'number': 'Number',
    'business_service': 'Service',
    'opened_at': 'Opened',
    'resolved_at': 'Resolved',
    'short_description': 'Short description',
    'state': 'State',
    'priority': 'Priority',
    'assignment_group': 'Assignment group',
    'assigned_to': 'Assigned to',
    'close_code': 'Close code',
    'close_notes': 'Close notes',
    'u_eus_es_ict_service': 'ICT Service',
    'u_eus_es_ict_subservice': 'ICT Subservice',
    'reopen_count': 'Reopen count',
    'opened_by': 'Opened by',
    'description': 'Description',
}


def normalizar_colunas(df):
    renomear = {
        col: _MAPA_COLUNAS_SERVICENOW[col.strip().lower()]
        for col in df.columns
        if col.strip().lower() in _MAPA_COLUNAS_SERVICENOW
        and _MAPA_COLUNAS_SERVICENOW[col.strip().lower()] not in df.columns
    }
    return df.rename(columns=renomear)

def derivar_resolvidos(df):
    df_res = df.dropna(subset=['Resolved', 'SLA - Dias (8 h)']).copy()
    df_res['Mes_Resolved_Sort'] = df_res['Resolved'].dt.to_period('M')
    df_res['Mes_Display'] = df_res['Resolved'].apply(formatar_mes_pt)
    return df_res


def processar_base(caminho, dias_nao_uteis_extra=None):
    calc = CalculadoraSLA(dias_nao_uteis_extra=dias_nao_uteis_extra)
    try: df = pd.read_excel(caminho)
    except: df = pd.read_csv(caminho)

    df = normalizar_colunas(df)

    df['Opened'] = pd.to_datetime(df['Opened'], errors='coerce', dayfirst=True)
    df['Resolved'] = pd.to_datetime(df['Resolved'], errors='coerce', dayfirst=True)
    df['Empresa'] = df['Service'].apply(identificar_empresa)

    partes = df['Short description'].apply(extrair_partes_description)
    df['Macro'] = [x[0] for x in partes]
    df['Categoria'] = [x[1] for x in partes]
    df['SubCategoria'] = [x[2] for x in partes]
    df['Descricao_Tratada'] = [x[3] for x in partes]

    df['SLA - Dias (8 h)'] = df.apply(lambda r: calc.calcular_sla(r['Opened'], r['Resolved']), axis=1)
    df['TMA - Dias corridos'] = (df['Resolved'] - df['Opened']).dt.total_seconds() / 86400.0
    df.loc[df['TMA - Dias corridos'] < 0, 'TMA - Dias corridos'] = 0

    df_res = derivar_resolvidos(df)

    return df_res, df

File: web/lib/processador-dados.ts
import { CalculadoraSLA } from "./calculadora-sla";
import type { Incidente } from "./types";

const MESES_PT = ["Jan", "Fev", "Mar", "Abr", "Mai", "Jun", "Jul", "Ago", "Set", "Out", "Nov", "Dez"];

export function identificarEmpresa(serviceStr: unknown): string {
  if (typeof serviceStr !== "string") return "Outros";
  const s = serviceStr.toUpperCase();
  if (s.includes("CEARA") || s.includes("COELCE")) return "CE";
  if (s.includes("RIO") || s.includes("AMPLA")) return "RJ";
  if (s.includes("SAO PAULO") || s.includes("SP")) return "SP";
  return "Outros";
}

export function extrairPartesDescription(shortDesc: unknown): {
  macro: string;
  categoria: string;
  subcategoria: string;
  descLimpa: string;
} {
  if (typeof shortDesc !== "string") {
    return { macro: "N/A", categoria: "N/A", subcategoria: "N/A", descLimpa: "N/A" };
  }

  let macro = "N/A";
  let categoria = "N/A";
  let subcategoria = "N/A";
  let descLimpa = shortDesc;

  const macroMatch = shortDesc.match(/\[(.*?)\]/);
  if (macroMatch) {
    macro = macroMatch[1].trim().toUpperCase();
    descLimpa = descLimpa.replace(macroMatch[0], "");
  }

  const catMatch = descLimpa.match(/\((.*?)\)/);
  if (catMatch) {
    const partes = catMatch[1].split("/");
    categoria = partes[0] ? partes[0].trim().toUpperCase() : "N/A";
    subcategoria = partes[1] ? partes[1].trim().toUpperCase() : "N/A";
    descLimpa = descLimpa.replace(catMatch[0], "");
  }

  descLimpa = descLimpa.replace(/\s+/g, " ").trim();
  if (!descLimpa) descLimpa = "N/A";

  return { macro, categoria, subcategoria, descLimpa };
}

export function formatarMesPt(data: Date | null): string {
  if (!data || isNaN(data.getTime())) return "N/A";
  return `${MESES_PT[data.getUTCMonth()]}/${String(data.getUTCFullYear()).slice(2)}`;
}

const MAPA_COLUNAS_SERVICENOW: Record<string, string> = {
  number: "Number",
  business_service: "Service",
  opened_at: "Opened",
  resolved_at: "Resolved",
  short_description: "Short description",
  state: "State",
  priority: "Priority",
  assignment_group: "Assignment group",
  assigned_to: "Assigned to",
  close_code: "Close code",
  close_notes: "Close notes",
  u_eus_es_ict_service: "ICT Service",
  u_eus_es_ict_subservice: "ICT Subservice",
  reopen_count: "Reopen count",
  opened_by: "Opened by",
  description: "Description",
};

export function normalizarColunas(rows: Record<string, unknown>[]): Record<string, unknown>[] {
  if (rows.length === 0) return rows;
  const colunasOriginais = Object.keys(rows[0]);
  const renomear: Record<string, string> = {};
  for (const col of colunasOriginais) {
    const alvo = MAPA_COLUNAS_SERVICENOW[col.trim().toLowerCase()];
    if (alvo && !colunasOriginais.includes(alvo)) {
      renomear[col] = alvo;
    }
  }
  if (Object.keys(renomear).length === 0) return rows;
  return rows.map((row) => {
    const novo: Record<string, unknown> = { ...row };
    for (const [origem, destino] of Object.entries(renomear)) {
      novo[destino] = novo[origem];
    }
    return novo;
  });
}

/** Parseia datas no padrão ServiceNow (dayfirst: DD/MM/YYYY HH:mm:ss ou variações ISO)
 * como horário "de parede" (sem conversão de fuso), igual ao pandas com tz naive. */
export function parseDataServiceNow(valor: unknown): Date | null {
  if (valor == null || valor === "") return null;
  if (valor instanceof Date) return isNaN(valor.getTime()) ? null : valor;

  const s = String(valor).trim();

  // dd/mm/yyyy[ hh:mm[:ss]]
  const brMatch = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})(?:[ T](\d{1,2}):(\d{2})(?::(\d{2}))?)?/);
  if (brMatch) {
    const [, dd, mm, yyyy, hh, min, ss] = brMatch;
    return new Date(Date.UTC(
      Number(yyyy), Number(mm) - 1, Number(dd),
      hh ? Number(hh) : 0, min ? Number(min) : 0, ss ? Number(ss) : 0
    ));
  }

  // yyyy-mm-dd[ hh:mm[:ss]]
  const isoMatch = s.match(/^(\d{4})-(\d{1,2})-(\d{1,2})(?:[ T](\d{1,2}):(\d{2})(?::(\d{2}))?)?/);
  if (isoMatch) {
    const [, yyyy, mm, dd, hh, min, ss] = isoMatch;
    return new Date(Date.UTC(
      Number(yyyy), Number(mm) - 1, Number(dd),
      hh ? Number(hh) : 0, min ? Number(min) : 0, ss ? Number(ss) : 0
    ));
  }

  const generico = new Date(s);
  return isNaN(generico.getTime()) ? null : generico;
}

export function processarBase(
  linhasBrutas: Record<string, unknown>[],
  diasNaoUteisExtra?: string[]
): Incidente[] {
  const calc = new CalculadoraSLA(diasNaoUteisExtra);
  const linhas = normalizarColunas(linhasBrutas);

  return linhas.map((row): Incidente => {
    const opened = parseDataServiceNow(row["Opened"]);
    const resolved = parseDataServiceNow(row["Resolved"]);
    const empresa = identificarEmpresa(row["Service"]);
    const { macro, categoria, subcategoria, descLimpa } = extrairPartesDescription(row["Short description"]);

    const sla = calc.calcularSla(opened, resolved);
    let tma: number | null = null;
    if (opened && resolved) {
      tma = (resolved.getTime() - opened.getTime()) / 86400000;
      if (tma < 0) tma = 0;
    }

    return {
      Number: String(row["Number"] ?? ""),
      Service: (row["Service"] as string) ?? null,
      Opened: opened ? opened.toISOString() : null,
      Resolved: resolved ? resolved.toISOString() : null,
      "Short description": (row["Short description"] as string) ?? null,
      State: (row["State"] as string) ?? null,
      Priority: (row["Priority"] as string) ?? null,
      "Assignment group": (row["Assignment group"] as string) ?? null,
      "ICT Service": (row["ICT Service"] as string) ?? null,
      "Assigned to": (row["Assigned to"] as string) ?? null,
      "Close code": (row["Close code"] as string) ?? null,
      "Close notes": (row["Close notes"] as string) ?? null,
      "Reopen count": row["Reopen count"] != null ? Number(row["Reopen count"]) : null,
      Empresa: empresa,
      Macro: macro,
      Categoria: categoria,
      SubCategoria: subcategoria,
      Descricao_Tratada: descLimpa,
      "SLA - Dias (8 h)": sla,
      "TMA - Dias corridos": tma,
    };
  });
}

export function derivarResolvidos(incidentes: Incidente[]): Incidente[] {
  return incidentes.filter((i) => i.Resolved != null && i["SLA - Dias (8 h)"] != null);
}

export function mesDisplay(iso: string | null): string {
  if (!iso) return "N/A";
  return formatarMesPt(new Date(iso));
}

export function mesSort(iso: string | null): string {
  if (!iso) return "9999-99";
  const d = new Date(iso);
  return `${d.getUTCFullYear()}-${String(d.getUTCMonth() + 1).padStart(2, "0")}`;
}
